### PR TITLE
[move-ide] Added smart auto-completion for colon-colon

### PR DESCRIPTION
**Pull request by awelc**
## Description 

Currently, this PR implements most of functionality for `::` auto-completion when the cursor is on one of the identifiers in the name access chains.

There are some missing parts here, arguably better split to separate PRs:
- handling of access chains in attributes
- handling of access chainss in `use` statements

Also, this PR actually removes `:` completion (auto-completion start when typing the first character after `:`) which is consistent with what rust-analyzer does. While we can implement auto-completion right after `:`, the question is whether we should

## Test plan 

All new and old tests must pass

### CODE REVIEW

File: external-crates/move/crates/move-compiler/src/shared/ide.rs [hunk 1]
```diff
@@ -82,11 +82,11 @@ pub struct DotAutocompleteInfo {
...
 pub struct AliasAutocompleteInfo {
     /// Numerical addresses that are valid autocompletes
-    pub addresses: BTreeSet<(Symbol, NumericalAddress)>,
+    pub addresses: BTreeMap<Symbol, NumericalAddress>,
     /// Modules that are valid autocompletes
-    pub modules: BTreeSet<(Symbol, E::ModuleIdent)>,
+    pub modules: BTreeMap<Symbol, E::ModuleIdent>,
     /// Members that are valid autocompletes
-    pub members: BTreeSet<(Symbol, E::ModuleIdent, Name)>,
+    pub members: BTreeMap<E::ModuleIdent, BTreeMap<Symbol, Name>>,
```
Comment (cgswords) [inline external-crates/move/crates/move-compiler/src/shared/ide.rs:89]: Is there a good reason to change these over? It doesn't matter too much to me, I'm just curious.

Comment (awelc) [reply external-crates/move/crates/move-compiler/src/shared/ide.rs]: At some point, I thought I'd need to search this by-module and figured it would be a better structure for this. I ended up not needing this at all and reverted this change

File: external-crates/move/crates/move-analyzer/src/completion.rs [hunk 2]
```diff
@@ -361,6 +389,640 @@ fn dot(symbols: &Symbols, use_fpath: &Path, position: &Position) -> Vec<Completi
...
+    path_index: usize,
+    colon_colon_triggered: bool,
+    completions: &mut Vec<CompletionItem>,
+) {
+    let ChainComponentInfo {
+        loc: prev_loc,
+        kind: prev_kind,
+    } = prev_info;
+
+    let mut at_colon_colon = false;
```
Comment (cgswords) [inline external-crates/move/crates/move-analyzer/src/completion.rs:723]: Please split this up into a few smaller functions, maybe something like this:

```suggestion
/// Walks down a name chain, looking for the relevant portion that contains the cursor. When. it finds, it calls to `name_chain_entry_completions` to compute and return the completions.
fn completions_for_name_chain_entry(
    symbols: &Symbols,
    cursor: &CursorContext,
    info: &AliasAutocompleteInfo,
    prev_info: ChainComponentInfo,
    chain_target: ChainCompletionTarget,
    path_entries: &[Name],
    path_index: usize,
    colon_colon_triggered: bool,
) -> Vec<CompletionItem> {
    let ChainComponentInfo {
        loc: prev_loc,
        kind: prev_kind,
    } = prev_info;

    let at_colon_colon =  cursor_at_colon_colon(path_entries, path_index, cursor);
    // we are at `::`, or at some component's identifier
    if at_colon_colon || path_entries[path_index].loc.contains(&cursor.loc) {
        name_chain_entry_completions(
            symbols, cursor, info, ..., prev_kind
        )
    } else if let Some(next_kind) = next_name_chain_kind(..., prev_kind) {
            entries_completions(
                symbols,
                cursor,
                info,
                ChainComponentInfo::new(component_name.loc, next_kind),
                chain_target,
                path_entries,
                path_index + 1,
                colon_colon_triggered,
            )
    } else {
        vec![]
    }
```

You could also consider doing it in two steps, so replace the call to `entries_completions` with:

```
if let Some(completion_context_info) = find_completion_info(symbols, cursor, ...) {
  completions.extend(name_chain_entry_completions(symbols, ..., completion_context_info));
}
```

File: external-crates/move/crates/move-analyzer/src/completion.rs [hunk 3]
```diff
@@ -361,6 +389,640 @@ fn dot(symbols: &Symbols, use_fpath: &Path, position: &Position) -> Vec<Completi
...
+        match mod_ident.address {
+            Address::Numerical {
+                name: _,
+                value,
+                name_conflict: _,
+            } if value.value == pkg_addr => return true,
+            _ => (),
+        }
+    }
+    false
```
Comment (cgswords) [inline external-crates/move/crates/move-analyzer/src/completion.rs:824]:
```suggestion
symbols.file_mods.values()
    .flatten()
    .any(|mdef| match mdef.ident.address {
        Address::Numerical { value, .. } if value.value == pkg_addr => true,
        _ => false,
    })
```

File: external-crates/move/crates/move-analyzer/src/completion.rs [hunk 4]
```diff
@@ -361,6 +389,640 @@ fn dot(symbols: &Symbols, use_fpath: &Path, position: &Position) -> Vec<Completi
...
+            Address::NamedUnassigned(name) if name == pkg_name => return true,
+            Address::Numerical {
+                name,
+                value: _,
+                name_conflict: _,
+            } if name == Some(pkg_name) => return true,
+            _ => (),
+        }
+    }
+    false
```
Comment (cgswords) [inline external-crates/move/crates/move-analyzer/src/completion.rs:850]:
```suggestion
symbols.file_mods.values()
    .flatten()
    .map(|mdef| &mdef.ident)
    .any(|mod_ident| match &mod_ident.address {
        Address::NamedUnassigned(name) if name == &pkg_name => true,
        Address::Numerical { name: Some(name), .. } if name == &pkg_name => true,
        _ => false,
    })
```

File: external-crates/move/crates/move-analyzer/src/completion.rs [hunk 5]
```diff
@@ -361,6 +389,640 @@ fn dot(symbols: &Symbols, use_fpath: &Path, position: &Position) -> Vec<Completi
...
+                if let Some(n) = name {
+                    addresses.insert(n.to_string());
+                }
+                addresses.insert(value.to_string());
+            }
+            Address::NamedUnassigned(n) => {
+                addresses.insert(n.to_string());
+            }
+        }
+    }
```
Comment (cgswords) [inline external-crates/move/crates/move-analyzer/src/completion.rs:884]:
```suggestion
symbols.file_mods.values()
    .flatten()
    .map(|mdef| &mdef.ident)
    .for_each(|mod_ident| {
        match &mod_ident.address {
            Address::Numerical { name, value, .. } => {
                if let Some(n) = name {
                    addresses.insert(n.to_string());
                }
                addresses.insert(value.to_string());
            }
            Address::NamedUnassigned(n) => {
                addresses.insert(n.to_string());
            }
        }
    });
```

File: external-crates/move/crates/move-analyzer/src/completion.rs [hunk 6]
```diff
@@ -361,6 +389,640 @@ fn dot(symbols: &Symbols, use_fpath: &Path, position: &Position) -> Vec<Completi
...
+
+    addresses
+}
+
+/// Handle path auto-completion at a given position. The gist of this approach is to first identify
+/// what the first component of the access chain represents (as it may be a package, module or a
+/// member) and if the chain has other components, recursively process them in turn to either
+/// - finish auto-completion if cursor is on a given component's identifier
+/// - identify what the subsequent component represents and keep going
+fn path_completions(
```
Comment (cgswords) [inline external-crates/move/crates/move-analyzer/src/completion.rs:894]:
```suggestion
/// Handle name chain auto-completion at a given position. The gist of this approach is to first identify
/// what the first component of the access chain represents (as it may be a package, module or a
/// member) and if the chain has other components, recursively process them in turn to either
/// - finish auto-completion if cursor is on a given component's identifier
/// - identify what the subsequent component represents and keep going
fn name_chain_completions(
```

File: external-crates/move/crates/move-analyzer/src/completion.rs [hunk 7]
```diff
@@ -361,6 +389,640 @@ fn dot(symbols: &Symbols, use_fpath: &Path, position: &Position) -> Vec<Completi
...
+    symbols: &Symbols,
+    cursor: &CursorContext,
+    colon_colon_triggered: bool,
+) -> (Vec<CompletionItem>, bool) {
+    eprintln!("looking for colon(s)");
+    let mut completions = vec![];
+    let mut only_custom_items = false;
```
Comment (cgswords) [inline external-crates/move/crates/move-analyzer/src/completion.rs:901]: We don't need this binding, really.

File: external-crates/move/crates/move-analyzer/src/completion.rs [hunk 8]
```diff
@@ -361,6 +389,640 @@ fn dot(symbols: &Symbols, use_fpath: &Path, position: &Position) -> Vec<Completi
...
+    if leading_name.loc.contains(&cursor.loc) {
+        // at first position of the chain suggest all packages that are available regardless of what
+        // the leading name represents, as a package always fits at that position, for example:
+        // OxCAFE::...
+        // some_name::...
+        // ::some_name
+        //
+        for n in all_packages(symbols, &info) {
+            completions.push(completion_item(n.as_str(), CompletionItemKind::UNIT));
+        }
```
Comment (cgswords) [inline external-crates/move/crates/move-analyzer/src/completion.rs:942]:
```suggestion
completions.extend(
    all_packages(symbols, &info)
        .map(|n| completion_item(n.as_str(), CompletionItemKind::UNIT))
);
```

File: external-crates/move/crates/move-analyzer/src/completion.rs [hunk 9]
```diff
@@ -361,6 +389,640 @@ fn dot(symbols: &Symbols, use_fpath: &Path, position: &Position) -> Vec<Completi
...
+            ));
+            if matches!(chain_target, ChainCompletionTarget::Type) {
+                completions.extend(primitive_types());
+                info.type_params.iter().for_each(|t| {
+                    completions.push(completion_item(
+                        t.as_str(),
+                        CompletionItemKind::TYPE_PARAMETER,
+                    ))
+                });
+            }
```
Comment (cgswords) [inline external-crates/move/crates/move-analyzer/src/completion.rs:963]:
```suggestion
    completions.extend(
        info.modules.iter().map(|(n, _)| completion_item(n.as_str(), CompletionItemKind::MODULE))
    );

    completions.extend(all_first_position_member_completions(
        symbols,
        &info.members,
        chain_target,
    ));

    if matches!(chain_target, ChainCompletionTarget::Type) {
        completions.extend(primitive_types());
        completions.extend(
            info.type_params.iter().map(|t| completion_item(t.as_str(), CompletionItemKind::TYPE_PARAMETER))
        );
    }
```

File: external-crates/move/crates/move-analyzer/src/completion.rs [hunk 10]
```diff
@@ -361,6 +389,640 @@ fn dot(symbols: &Symbols, use_fpath: &Path, position: &Position) -> Vec<Completi
...
+            LeadingNameAccess_::GlobalAddress(n) => {
+                // if leading name is global address then the first component can only be a
+                // package
+                if is_package_name(symbols, &info, n) {
+                    Some(ChainComponentKind::Package(leading_name))
+                } else {
+                    None
+                }
+            }
+        };
```
Comment (cgswords) [inline external-crates/move/crates/move-analyzer/src/completion.rs:1004]: Make this a function

File: external-crates/move/crates/move-compiler/src/shared/ide.rs [hunk 11]
```diff
@@ -82,9 +82,9 @@ pub struct DotAutocompleteInfo {
 #[derive(Default, Debug, Clone)]
 pub struct AliasAutocompleteInfo {
     /// Numerical addresses that are valid autocompletes
-    pub addresses: BTreeSet<(Symbol, NumericalAddress)>,
+    pub addresses: BTreeMap<Symbol, NumericalAddress>,
```
Comment (cgswords) [inline external-crates/move/crates/move-compiler/src/shared/ide.rs:85]: What is the motivation for this change?

**Review by cgswords — CHANGES_REQUESTED** [review]
Mostly LGTM, just loop / style cleanup nits. I did make a few points where I think the code should be a little more chunked up for easier readability / maintenance in the future.

File: external-crates/move/crates/move-analyzer/src/symbols.rs [hunk 12]
```diff
@@ -394,6 +394,13 @@ pub struct CursorContext {
     pub loc: Loc,
 }
 
+#[derive(Clone, Debug, Copy)]
+pub enum ChainCompletionTarget {
```
[the remark this answers is not in the document]
Comment (awelc) [reply external-crates/move/crates/move-analyzer/src/symbols.rs:398]: Happy to change if from `Target`, though it does not sound that bad to me as we are indeed aiming for a function or a type with the completed chain. Also `Context` sounds a bit confusing to me in turn, as it suggests something that surrounds the chain rather than what the chain represents. Not sure what the best compromise here could be (`ChainEntityType/Kind`?)

File: external-crates/move/crates/move-compiler/src/shared/ide.rs (hunk 11 above)
Comment (awelc) [reply external-crates/move/crates/move-compiler/src/shared/ide.rs:85]: Faster look up on the symbol key. Not a big issue at all (and not a big gain :-) )  - can certainly revert if that's preferrable

File: external-crates/move/crates/move-analyzer/src/completion.rs (hunk 7 above)
Comment (awelc) [reply external-crates/move/crates/move-analyzer/src/completion.rs:958]: Not really, but it's a piece of documentation as well

File: external-crates/move/crates/move-analyzer/src/completion.rs [hunk 13]
```diff
@@ -361,6 +389,640 @@ fn dot(symbols: &Symbols, use_fpath: &Path, position: &Position) -> Vec<Completi
...
+    if !at_colon_colon {
+        // we are not at the last `::` but we may be at an intermediate one
+        if colon_colon_triggered
+            && path_index < path_entries.len()
+            && cursor.loc.start() > prev_loc.end()
+            && cursor.loc.end() <= path_entries[path_index].loc.start()
+        {
+            at_colon_colon = true;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (awelc) [reply external-crates/move/crates/move-analyzer/src/completion.rs:805]: Because we first need to check if we are at the last `::` (and return if we are not) and only then check if we at an intermediate one, to me it feels like pulling this into its own function would result in code that's more confusing

File: external-crates/move/crates/move-analyzer/src/completion.rs [hunk 14]
```diff
@@ -361,6 +389,640 @@ fn dot(symbols: &Symbols, use_fpath: &Path, position: &Position) -> Vec<Completi
...
+            return;
+        }
+    }
+
+    if !at_colon_colon {
+        // we are not at the last `::` but we may be at an intermediate one
+        if colon_colon_triggered
+            && path_index < path_entries.len()
+            && cursor.loc.start() > prev_loc.end()
+            && cursor.loc.end() <= path_entries[path_index].loc.start()
```
[the remark this answers is not in the document]
Comment (awelc) [reply external-crates/move/crates/move-analyzer/src/completion.rs:801]: It checks if we are after ending location of one name (or leading name) and before the start of the other one whereas `path_entries[path_index].loc.contains(&cursor.loc)` checks if we are at a given name/identifier.

File: external-crates/move/crates/move-analyzer/src/completion.rs [hunk 15]
```diff
@@ -361,6 +389,640 @@ fn dot(symbols: &Symbols, use_fpath: &Path, position: &Position) -> Vec<Completi
...
+
+/// Returns completion item if a given name/alias identifies a valid member of a given module
+/// available in the completion scope
+fn first_position_member_completion(
+    symbols: &Symbols,
+    mod_ident: &ModuleIdent_,
+    member_alias: &Symbol,
+    member_name: &Symbol,
+    chain_target: ChainCompletionTarget,
+) -> Option<CompletionItem> {
```
[the remark this answers is not in the document]
Comment (awelc) [reply external-crates/move/crates/move-analyzer/src/completion.rs:501]: Because one is dealing with all available completions and the other checks if a given completion is valid, I don't really see an easy and clean way to share the code without resorting to constructs that I am not sure would actually be an improvement here (e.g., callbacks or macros)

File: external-crates/move/crates/move-analyzer/src/completion.rs [hunk 16]
```diff
@@ -55,7 +77,7 @@ fn completion_item(label: &str, kind: CompletionItemKind) -> CompletionItem {
 /// all specification language keywords, but in the future it should be modified to only do so
 /// within a spec block.
 fn keywords() -> Vec<CompletionItem> {
```
[the remark this answers is not in the document]
Comment (cgswords) [reply external-crates/move/crates/move-analyzer/src/completion.rs]: I do!

File: external-crates/move/crates/move-analyzer/src/symbols.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (cgswords) [reply external-crates/move/crates/move-analyzer/src/symbols.rs:398]: I like `kind` a lot, and it's consistent with our verbage in the compiler.

File: external-crates/move/crates/move-analyzer/src/completion.rs [hunk 17]
```diff
@@ -361,6 +389,640 @@ fn dot(symbols: &Symbols, use_fpath: &Path, position: &Position) -> Vec<Completi
     completions
 }
 
+/// Returns completion items for all members of a given module for n-th access chain component
+/// position where n > 1
+fn all_n_position_member_completions(
```
[the remark this answers is not in the document]
Comment (cgswords) [reply external-crates/move/crates/move-analyzer/src/completion.rs]: This function name with `all_n` is deeply uncearl to me. Even with the doc comment, it's strange. It's also incorrect as it does return `all`; it's gated by the expected kind. `nth_position_completions` would be better, and `access_cahin_nth_position_completions` would be even better to separate it from "dot paths".

File: external-crates/move/crates/move-analyzer/src/completion.rs [hunk 18]
```diff
@@ -361,6 +389,640 @@ fn dot(symbols: &Symbols, use_fpath: &Path, position: &Position) -> Vec<Completi
...
+            completions.push(completion_item(cname, CompletionItemKind::CONSTANT));
+        }
+    }
+
+    completions
+}
+
+/// Returns completion item if a given name/alias identifies a valid member of a given module
+/// available in the completion scope
+fn first_position_member_completion(
```
[the remark this answers is not in the document]
Comment (cgswords) [reply external-crates/move/crates/move-analyzer/src/completion.rs]: Same comment on names

File: external-crates/move/crates/move-analyzer/src/completion.rs [hunk 19]
```diff
@@ -361,6 +389,640 @@ fn dot(symbols: &Symbols, use_fpath: &Path, position: &Position) -> Vec<Completi
...
+        };
+
+        completions.push(CompletionItem {
+            label,
+            kind: Some(CompletionItemKind::ENUM_MEMBER),
+            insert_text: Some(insert_text),
+            insert_text_format: Some(InsertTextFormat::SNIPPET),
+            ..Default::default()
+        });
+    }
```
[the remark this answers is not in the document]
Comment (awelc) [reply external-crates/move/crates/move-analyzer/src/completion.rs]: Finessed using `filter_map`

**Review by cgswords — APPROVED** [review]
LGTM. Great work, and exciting!
